metrics/fidelity/utility.py

```diff
@@ -0,0 +1,174 @@
+    :param model: pytorch nn Module
+    :param attention_masks: tensor of attention masks
+    :return: numpy array of labels and probabilities associated with that label
+    """
+    if attention_masks is not None:
+        result = model.forward(input_ids=input_ids.to(model.device), attention_mask=attention_masks.to(model.device))
+    else:
+        result = model.forward(input_ids=input_ids.to(model.device))
+
+    if model.device == "cuda":
+        result["py_index"] = result["py_index"].detach().cpu()
+        result["probs"] = result["probs"].detach().cpu()
+
+    return result["py_index"].numpy(), result["probs"].numpy()
+
+
+def compute_null_diff(input_ids, model, predictions, prob_y_hat, tokenizer):
+    """
+    Calculate the null difference which helps in accounting for model-dependent baseline performance
+    null difference = max(0, p(y_hat/x) - p(y_hat/x, 0))
+    returns the label and prediction probability
+    :param prob_y_hat: P(y_hat/x)
+    :param predictions: y_hat
+    :param input_ids: tensor of input ids
+    :param model: pytorch nn Module
+    :param tokenizer: data tokenizer
+    :return: null difference
+    """
+    input_ids_reduced, attention_mask_reduced = reduce(input_ids=input_ids, rationale=None, tokenizer=tokenizer)
+
+    predictions_0, prob_y_hat_0 = compute_predictions(
+        input_ids=input_ids_reduced, model=model, attention_masks=attention_mask_reduced
+    )
+    prob_y_hat_0_predicted_class = prob_y_hat_0[np.arange(len(prob_y_hat_0)), predictions]
+
+    null_difference = 1 - compute_fidelity(
+        prob_y_hat=prob_y_hat, prob_y_hat_alpha=prob_y_hat_0_predicted_class, fidelity_type="sufficiency"
+    )
+
+    return null_difference
+
+
+def reduce(input_ids, rationale=None, tokenizer=None, fidelity_type="sufficiency"):
+    """
+    Reduce the input_ids based on the alpha values or the rationale values
+    sufficiency - reduce by keeping only the tokens which are in the rationale
+    comprehensiveness - reduce by keeping all the tokens which are not in the rationale
+    :param input_ids: tensor of input ids
+    :param rationale: a list of 0s or 1s corresponding to a token's presence in the rationale
+    :param tokenizer: data tokenizer
+    :param fidelity_type: sufficiency or comprehensiveness
+    :return: tensor of reduced input ids and reduced attention masks
+    """
+    input_ids_reduced = []
+    attention_mask_reduced = []
+    for idx in range(len(input_ids)):
+        if rationale is None:
+            rationale_i = [0] * len(input_ids[idx])
+        else:
+            rationale_i = rationale[idx]
+        input_ids_reduced_i = []
+        attention_mask_reduced_i = []
+        for j in range(len(input_ids[idx])):
+            if input_ids[idx][j] in [0, 2]:
+                input_ids_reduced_i.append(input_ids[idx][j].item())
+                attention_mask_reduced_i.append(1)
+            else:
+                if fidelity_type == "sufficiency" and rationale_i[j] >= 0.5:
```

Comment:
maybe we should make 0.5 a parameter?

Reply:
You can call it "binarization_threshold"